chemprop/uncertainty/calibrator.py

```diff
@@ -78,58 +81,251 @@ def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
+        in Multiclass and Multilabel Prediction." arXiv Preprint 2020, https://arxiv.org/abs/2004.10181
+    """
+
+    def __init__(self, alpha: float):
+        super().__init__()
+        self.alpha = alpha
+        if not 0 <= self.alpha <= 1:
+            raise ValueError(
+                f"The error rate (i.e., alpha) should be between 0 and 1. Got {self.alpha}."
+            )
+
+    @staticmethod
+    def nonconformity_scores(preds: Tensor):
+        r"""
+        Compute nonconformity score as the negative of the predicted probability.
+
+        .. math::
+            s_i = -\hat{f}(X_i)_{Y_i}
+        """
+        return -preds
+
     def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> None:
-        ...
-        return
+        if targets.shape[1] < 2:
+            raise ValueError(
+                f"The number of tasks should be laerger than 1. Got {targets.shape[1]}."
+            )
+
+        has_zeros = torch.any(targets == 0, dim=1)
+        index_zeros = targets[has_zeros] == 0
+        scores_in = self.nonconformity_scores(uncs[has_zeros])
+        masked_scores_in = scores_in * index_zeros.float() + torch.where(
+            index_zeros, torch.zeros_like(scores_in), torch.tensor(float("inf"))
+        )
+        calibration_scores_in = torch.min(
+            masked_scores_in.masked_fill(~mask, float("inf")), dim=1
+        ).values
+
+        has_ones = torch.any(targets == 1, dim=1)
+        index_ones = targets[has_ones] == 1
+        scores_out = self.nonconformity_scores(uncs[has_ones])
+        masked_scores_out = scores_out * index_ones.float() + torch.where(
+            index_ones, torch.zeros_like(scores_out), torch.tensor(float("-inf"))
+        )
+        calibration_scores_out = torch.max(
+            masked_scores_out.masked_fill(~mask, float("-inf")), dim=1
+        ).values
+
+        self.tout = torch.quantile(
+            calibration_scores_out, 1 - self.alpha / 2, interpolation="higher"
+        )
+        self.tin = torch.quantile(calibration_scores_in, self.alpha / 2, interpolation="higher")
 
     def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        ...
-        return
+        scores = self.nonconformity_scores(uncs)
+
+        cal_preds_in = (scores <= self.tin).int()
+        cal_preds_out = (scores <= self.tout).int()
+        cal_preds_in_out = torch.cat((cal_preds_in, cal_preds_out), dim=1)
+
+        return preds, cal_preds_in_out
 
 
 @UncertaintyCalibratorRegistry.register("conformal-multiclass")
-class ConformalMulticlassCalibrator(UncertaintyCalibrator):
+class MulticlassConformalCalibrator(UncertaintyCalibrator):
+    r"""Create a prediction sets of possible labels :math:`C(X_{\text{test}}) \subset \{1, \ldots, K\}` that follows:
+
+    .. math::
+        1 - \alpha \leq \mathbb{P}(Y_{\text{test}} \in C(X_{\text{test}})) \leq 1 - \alpha + \frac{1}{n + 1}
```

Comment:
```suggestion
        1 - \alpha \leq \Pr (Y_{\text{test}} \in C(X_{\text{test}})) \leq 1 - \alpha + \frac{1}{n + 1}
```

Reply:
@davidegraff is this the format you think of?
![Screen Shot 2024-08-27 at 1 33 52 PM](https://github.com/user-attachments/assets/e57f9c2c-c1a2-494e-b771-3ed5063d69fa)

Reply:
yep!